src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp

```diff
@@ -516,59 +515,61 @@ extern "C" bool PalDetachThread(void* thread)
 
 #if !defined(USE_PORTABLE_HELPERS) && !defined(FEATURE_RX_THUNKS)
 
-#ifdef TARGET_APPLE
-static const struct section_64 *thunks_section;
-static const struct section_64 *thunks_data_section;
-#endif
-
 REDHAWK_PALEXPORT UInt32_BOOL REDHAWK_PALAPI PalAllocateThunksFromTemplate(HANDLE hTemplateModule, uint32_t templateRva, size_t templateSize, void** newThunksOut)
 {
 #ifdef TARGET_APPLE
-    int f;
-    Dl_info info;
+    vm_address_t addr, taddr;
+    vm_prot_t prot, max_prot;
+    kern_return_t ret;
 
-    int st = dladdr((const void*)hTemplateModule, &info);
-    if (st == 0)
+    // Allocate two contiguous ranges of memory: the first range will contain the trampolines
+    // and the second range will contain their data.
+    do
     {
-        return UInt32_FALSE;
-    }
+        ret = vm_allocate(mach_task_self(), &addr, templateSize * 2, VM_FLAGS_ANYWHERE);
+    } while (ret == KERN_ABORTED);
 
-    f = open(info.dli_fname, O_RDONLY);
-    if (f < 0)
+    if (ret != KERN_SUCCESS)
     {
         return UInt32_FALSE;
     }
 
-    // NOTE: We ignore templateRva since we would need to convert it to file offset
-    // and templateSize is useless too. Instead we read the sections from the
-    // executable and determine the size from them.
-    if (thunks_section == NULL)
+    do
+    {
+        ret = vm_remap(
+            mach_task_self(), &addr, templateSize, 0, VM_FLAGS_FIXED | VM_FLAGS_OVERWRITE,
```

Comment:
Just trying to understand the difference here, Mono seems to pass only `FALSE` in place of flags when calling `vm_remap`, why are we passing `VM_FLAGS_FIXED | VM_FLAGS_OVERWRITE` instead (or vice versa should we use the same on Mono side as well)?

Reply:
We should use this on Mono side as well. Mono uses a loop with vm_allocate(2*size)+vm_deallocate(1*size)+vm_remap. There's a race condition between the vm_deallocate+vm_remap not ending up at the same address. If the race condition is hit then it starts again from the beginning. `VM_FLAGS_OVERWRITE` gets rid of this race condition by skipping the intermediate reallocation.

Reply:
/cc: @lambdageek ^

Reply:
Yea seems like a good idea (I'm not intimately familiar with our infinite trampolines code, but it seems workable).

So the new loop would be:
- `vm_allocate(size)` returning `addr`
- set `tpage = addr+size`
- `vm_remap(&tpage, VM_FLAGS_FIXED | VM_FLAGS_OVERWRITE)`
- if success, we're done; if failure, `vm_deallocate(addr)` and repeat

?

Reply:
Yep